pkg/conn/baseconn.go

```diff
@@ -11,61 +11,84 @@
+//   Currently we have some types of connections exist
+//   Syncer:
+// 	  Worker Connection:
+//      DML connection:
+// 			execute some DML on Downstream DB, one unit has `syncer.WorkerCount` worker connections
+//  	DDL Connection:
+// 			execute some DDL on Downstream DB, one unit has one connection
+// 	  CheckPoint Connection:
+// 		interact with CheckPoint DB, one unit has one connection
+//    OnlineDDL connection:
+// 		interact with Online DDL DB, one unit has one connection
+//    ShardGroupKeeper connection:
+// 		interact with ShardGroupKeeper DB, one unit has one connection
+//
+//   Loader:
+// 	  Worker Connection:
+// 		execute some DML to Downstream DB, one unit has `loader.PoolSize` worker connections
+// 	  CheckPoint Connection:
+// 		interact with CheckPoint DB, one unit has one connection
+// 	  Restore Connection:
+// 		only use to create schema and table in restoreData,
+// 		it ignore already exists error and it should be removed after use, one unit has one connection
+//
+// each connection should have ability to retry on some common errors (e.g. tmysql.ErrTiKVServerTimeout) or maybe some specify errors in the future
+// and each connection also should have ability to reset itself during some specify connection error (e.g. driver.ErrBadConn)
 type BaseConn struct {
-	DB *sql.DB
-
-	// for reset
-	DSN string
+	DBConn *sql.Conn
 
 	RetryStrategy retry.Strategy
 
-	RawDBCfg *RawDBConfig
-}
-
-// RawDBConfig contains some low level database config
-type RawDBConfig struct {
-	MaxIdleConns int
+	// for reset
+	parentDB *BaseDB
 }
 
-// DefaultRawDBConfig returns a default raw database config
-func DefaultRawDBConfig() *RawDBConfig {
-	return &RawDBConfig{
-		MaxIdleConns: 2,
+// NewBaseConn builds BaseConn to connect real DB
+func NewBaseConn(conn *sql.Conn, strategy retry.Strategy, db *BaseDB) *BaseConn {
+	if strategy == nil {
+		strategy = &retry.FiniteRetryStrategy{}
 	}
+	return &BaseConn{conn, strategy, db}
 }
 
-// NewBaseConn builds BaseConn to connect real DB
-func NewBaseConn(dbDSN string, strategy retry.Strategy, rawDBCfg *RawDBConfig) (*BaseConn, error) {
-	db, err := sql.Open("mysql", dbDSN)
+// Reset resets this BaseConn, close old one and generate new *sql.Conn from its BaseDB
+func (conn *BaseConn) Reset(tctx *tcontext.Context) error {
+	err := conn.parentDB.CloseBaseConn(conn)
 	if err != nil {
-		return nil, terror.ErrDBDriverError.Delegate(err)
+		tctx.L().Warn("close connection during reset", log.ShortError(err))
 	}
-	// set max idle connection limit before any database call
-	db.SetMaxIdleConns(rawDBCfg.MaxIdleConns)
-	err = db.Ping()
+	baseConn, err := conn.parentDB.GetBaseConn(tctx.Context())
 	if err != nil {
-		db.Close()
-		return nil, terror.ErrDBDriverError.Delegate(err)
+		return err
 	}
-	if strategy == nil {
-		strategy = &retry.FiniteRetryStrategy{}
-	}
-	return &BaseConn{db, dbDSN, strategy, rawDBCfg}, nil
+	conn.DBConn = baseConn.DBConn
```

Comment:
two `BaseConn` instances share one `*sql.Conn` instance?

and `conn` not in `parentDB` anymore but `baseConn` in it now?